### PR TITLE
Optimize initial writes

### PR DESCRIPTION
**Pull request by geemus**
possible fix for #556

### CODE REVIEW

**Comment by nelhage** [comment]
I did the `read_nonblock` thing because I was worried about a case where we might want to send the request immediately, but don't yet have the body available. I can't actually think of a case in practice where this would be an issue, but if someone were depending on sending the headers immediately (some kind of weird streaming protocol?) this change might cause a regression. I'm happy to declare that use case too weird to support (and users could always use `:request_block` to support sending the headers and then sending the body after a delay).

**Comment by nelhage** [comment]
Otherwise, this should fix the issue so I'm fine with it! I confirmed it works in my test harness.

**Comment by geemus** [comment]
It's not super obvious, but that read actually will be non-block if the settings have non-block turned on (and blocking if it is turned off). So I think it should still support the case you mentioned. You can see where that fits into things here: https://github.com/excon/excon/blob/master/lib/excon/socket.rb#L31 Does that make sense/help?

For the test harness you mention, do you mean it shows that we now avoid the nagel/etc delay? Or just that it works in the general case? I'd be curious if you have ideas on how we might try to protect ourself from re-introducing this in the future.

Thanks!

**Comment by nelhage** [comment]
Ah, awesome on the nonblocking indirection. I think that makes sense.

Yeah, I noticed this via some HTTP benchmarking I was doing, that showed the Nagle/delayed ACK delay; With this change it no longer exhibits the delay.

I haven't looked too deeply at the test suite, but we could imagine injecting a fake `Socket` implementation that records individual `write` calls, and verifying that there's only one for a small request; I'd hope that + a comment in the test explaining why this is important would suffice to prevent another regression.

**Comment by geemus** [comment]
Sounds good. I'll go ahead and bring this in, but if you have a chance to help around regression prevention/testing I'd certainly appreciate the help. Thanks!

File: lib/excon/connection.rb [hunk 1]
```diff
@@ -155,6 +155,8 @@ def request_call(datum)
                 break
               end
             end
+          elsif body.nil?
+            socket.write(request) # write out request + headers
           elsif !body.nil? # write out body
```
Comment (Roman2K) [inline lib/excon/connection.rb]: Shouldn't that become an `else`? :sweat_smile:

Comment (geemus) [inline lib/excon/connection.rb]: Shouldn't break anything as elsif, but yeah, else seems easier/clearer. I've fixed it here: https://github.com/excon/excon/commit/24602e2316f0b08734eff4ff72db81662cb26126

Comment (Roman2K) [inline lib/excon/connection.rb]: Nice

**Pull request merged**